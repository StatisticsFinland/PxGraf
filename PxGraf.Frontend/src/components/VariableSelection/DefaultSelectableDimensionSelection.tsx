import React from 'react';
import { Autocomplete, TextField } from '@mui/material';
import { UiLanguageContext } from 'contexts/uiLanguageContext';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';
import { IDimensionValue } from 'types/cubeMeta';
import { EditorContext } from 'contexts/editorContext';

interface IDefaultSelectableDimensionSelection {
    dimensionCode: string;
    resolvedDimensionValueCodes: string[];
    options: IDimensionValue[],
}

const StyledAutocomplete = styled(Autocomplete)`
    flex-basis: 0;
    flex-grow: 1;
    width: 90%;
`;


export const DefaultSelectableDimensionSelection: React.FC<IDefaultSelectableDimensionSelection> = ({ dimensionCode, resolvedDimensionValueCodes, options }) => {
    const { language } = React.useContext(UiLanguageContext);
    const { defaultSelectables, setDefaultSelectables } = React.useContext(EditorContext);
    const { t } = useTranslation();

    // Note: array filtering to support selecting multiple values in the future
    const defaultOptionValues = defaultSelectables?.[dimensionCode] ? options.filter((option) => defaultSelectables[dimensionCode].includes(option.code)) : [];
    const [value, setValue] = React.useState<IDimensionValue>(defaultOptionValues[0] ? defaultOptionValues[0] : null);

    React.useEffect(() => {
        if (value && resolvedDimensionValueCodes.length > 0) {
            if (!resolvedDimensionValueCodes.includes(value.code)) {
                const defaultSelectablesCopy = { ...defaultSelectables };
                delete defaultSelectablesCopy[dimensionCode];
                setDefaultSelectables(defaultSelectablesCopy);
                setValue(null);
            }
        }
    }, [resolvedDimensionValueCodes]);

    const handleChange = (_evt, value: IDimensionValue) => {
        if (value) {
            setDefaultSelectables({
                ...defaultSelectables,
                [dimensionCode]: [value.code]
            })
        } else {
            const defaultSelectablesCopy = { ...defaultSelectables };
            delete defaultSelectablesCopy[dimensionCode];
            setDefaultSelectables(defaultSelectablesCopy);
        }
        setValue(value);
    };

    return (<StyledAutocomplete
        options={options.filter(option => resolvedDimensionValueCodes.indexOf(option.code) > -1)}
        getOptionLabel={(option: IDimensionValue) => option?.name[language] ?? option.code}
        isOptionEqualToValue={() => true}
        value={value}
        onChange={handleChange}
        noOptionsText={t("selectable.noSelections")}
        openText={t("selectable.open")}
        closeText={t("selectable.close")}
        clearText={t("selectable.clear")}
        renderInput={(params) => (
            <TextField {...params} label={t("selectable.labelText")} />
        )}
    />);

}

export default DefaultSelectableDimensionSelection;