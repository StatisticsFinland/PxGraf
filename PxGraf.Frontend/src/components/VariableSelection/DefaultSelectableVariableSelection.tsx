import React from 'react';
import { Autocomplete, TextField } from '@mui/material';
import { UiLanguageContext } from 'contexts/uiLanguageContext';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';
import { IDimensionValue } from 'types/cubeMeta';
import { EditorContext } from 'contexts/editorContext';

interface IDefaultSelectableVariableSelection {
    variableCode: string;
    resolvedVariableValueCodes: string[];
    options: IDimensionValue[],
}

const StyledAutocomplete = styled(Autocomplete)`
    flex-basis: 0;
    flex-grow: 1;
    width: 90%;
`;


export const DefaultSelectableVariableSelection: React.FC<IDefaultSelectableVariableSelection> = ({ variableCode, resolvedVariableValueCodes, options }) => {
    const { language } = React.useContext(UiLanguageContext);
    const { defaultSelectables, setDefaultSelectables } = React.useContext(EditorContext);
    const { t } = useTranslation();

    // Note: array filtering to support selecting multiple values in the future
    const defaultOptionValues = defaultSelectables?.[variableCode] ? options.filter((option) => defaultSelectables[variableCode].includes(option.code)) : [];
    const [value, setValue] = React.useState<IDimensionValue>(defaultOptionValues[0] ? defaultOptionValues[0] : null);

    React.useEffect(() => {
        if (value && resolvedVariableValueCodes.length > 0) {
            if (!resolvedVariableValueCodes.includes(value.code)) {
                const defaultSelectablesCopy = { ...defaultSelectables };
                delete defaultSelectablesCopy[variableCode];
                setDefaultSelectables(defaultSelectablesCopy);
                setValue(null);
            }
        }
    }, [resolvedVariableValueCodes]);

    const handleChange = (_evt, value: IDimensionValue) => {
        if (value) {
            setDefaultSelectables({
                ...defaultSelectables,
                [variableCode]: [value.code]
            })
        } else {
            const defaultSelectablesCopy = { ...defaultSelectables };
            delete defaultSelectablesCopy[variableCode];
            setDefaultSelectables(defaultSelectablesCopy);
        }
        setValue(value);
    };

    return (<StyledAutocomplete
        options={options.filter(option => resolvedVariableValueCodes.indexOf(option.code) > -1)}
        getOptionLabel={(option: IDimensionValue) => option?.name[language] ?? option.code}
        isOptionEqualToValue={() => true}
        value={value}
        onChange={handleChange}
        noOptionsText={t("selectable.noSelections")}
        openText={t("selectable.open")}
        closeText={t("selectable.close")}
        clearText={t("selectable.clear")}
        renderInput={(params) => (
            <TextField {...params} label={t("selectable.labelText")} />
        )}
    />);

}

export default DefaultSelectableVariableSelection;