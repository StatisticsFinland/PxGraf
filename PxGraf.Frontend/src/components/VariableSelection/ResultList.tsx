import React from 'react';
import { useTranslation } from 'react-i18next';
import { Stack, List, ListItem, ListItemText, Typography, Skeleton } from '@mui/material';
import styled from 'styled-components';
import { IDimensionValue } from 'types/cubeMeta';
import { UiLanguageContext } from 'contexts/uiLanguageContext';

interface IResultListProps {
    variableValues: IDimensionValue[]
    resolvedVariableValueCodes: string[]
}

const DenseList = styled(List)`
    width: 100%;
    border-style: solid;
    border-width: 1px;
    border-radius: 4px;
    position: relative;
    overflow: auto;
    max-height: 200px;
`;

const NoPaddingListItem = styled(ListItem)`
    padding-top: 0;
    padding-bottom: 0;
`;

const StyledEm = styled.em`
    color: #ccc;
`;

const StyledSkeleton = styled(Skeleton)`
    width: 50%;
`;

const DenseListWrapper = styled(Stack)`
    width: 100%;
`;

export const ResultList: React.FC<IResultListProps> = ({ variableValues, resolvedVariableValueCodes }) => {
    const { t } = useTranslation();
    const { uiContentLanguage } = React.useContext(UiLanguageContext);

    let listContent;
    if (resolvedVariableValueCodes == null) {
        listContent =
            <NoPaddingListItem>
                <StyledSkeleton variant="text" />
            </NoPaddingListItem>
    }
    else if (resolvedVariableValueCodes.length === 0) {
        listContent =
            <NoPaddingListItem>
                <StyledEm>{t("general.noResults")}</StyledEm>
            </NoPaddingListItem>
    }
    else {
        const resultTexts = resolvedVariableValueCodes.map(valueCode => {
            const value = variableValues.find(value => value.Code === valueCode);
            return value?.Name[uiContentLanguage] ?? valueCode;
        });

        listContent = resultTexts.map((value) =>
            <NoPaddingListItem key={value}>
                <ListItemText primary={value} />
            </NoPaddingListItem>
        );
    }

    return (
        <DenseListWrapper>
            <Typography>{t("general.results")}:</Typography>
            <DenseList dense
                sx={{
                    bgcolor: 'background.paper',
                    borderColor: 'divider',
                    '& ul': { padding: 0 },
                }}
            >
                {listContent}
            </DenseList>
        </DenseListWrapper>
    );
}

export default ResultList;