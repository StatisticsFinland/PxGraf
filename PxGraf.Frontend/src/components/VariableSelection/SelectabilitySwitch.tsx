import React from 'react';
import { useTranslation } from 'react-i18next';

import { FormControlLabel, FormControl, Switch } from '@mui/material';

interface ISelectabilitySwitchProps {
    selected: boolean,
    onChange: (newValue: boolean) => object
}

export const SelectabilitySwitch: React.FC<ISelectabilitySwitchProps> = ({ selected, onChange }) => {
    const { t } = useTranslation();

    return (
        <FormControl fullWidth>
            <FormControlLabel
                control={
                    <Switch
                        checked={selected ?? false}
                        onChange={(event) => onChange(event.target.checked)}
                    />
                }
                label={t("variableSelect.selectable")}
            />
        </FormControl>
    )
}

export default SelectabilitySwitch;