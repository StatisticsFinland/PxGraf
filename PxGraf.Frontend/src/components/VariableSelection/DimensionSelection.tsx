import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';

import { Stack, ListItemIcon, ListItemText, IconButton, MenuItem, Menu } from '@mui/material';

import MenuIcon from '@mui/icons-material/Menu';
import CheckIcon from '@mui/icons-material/Check';

import {
    getDefaultFilter, queryTypeLabels,
} from 'utils/dimensionSelectionHelpers';
import SelectabilitySwitch from './SelectabilitySwitch';
import ResultList from './ResultList';
import ManualPickDimensionSelection from './FilterComponents/ManualPickDimensionSelection';
import AllDimensionSelection from './FilterComponents/AllDimensionSelection';
import StartingFromDimensionSelection from './FilterComponents/StartingFromDimensionSelection';
import TopNDimensionSelection from './FilterComponents/TopNDimensionSelection';
import styled from 'styled-components';
import { IDimension } from 'types/cubeMeta';
import { FilterType, IDimensionQuery, Query } from 'types/query';
import DefaultSelectableDimensionSelection from './DefaultSelectableDimensionSelection';
import { EditorContext } from '../../contexts/editorContext';

interface IDimensionSelectionProps {
    dimension: IDimension
    resolvedDimensionValueCodes: string[]
    query: Query
}

const SelectorWrapper = styled(Stack)`
    width: 100%;
    align-items: flex-start;
`;

const ComponentWrapper = styled(Stack)`
    justify-content: space-between;
    align-items: flex-start;
`;

export const DimensionSelection: React.FC<IDimensionSelectionProps> = ({ dimension, resolvedDimensionValueCodes, query }) => {
    const [anchorEl, setAnchorEl] = useState(null);
    const { t } = useTranslation();
    const { setQuery } = React.useContext(EditorContext);
    const dimensionQuery = query[dimension.code];

    const onQueryChanged = (newQuery: IDimensionQuery) => {
        setQuery({
            ...query,
            [dimension.code]: newQuery
        });
    };

    const handleOpenMenuClick = (event) => {
        setAnchorEl(event.currentTarget);
    };

    const closeMenu = () => {
        setAnchorEl(null);
    };

    const handleMenuClick = (newFilterType: FilterType) => {
        closeMenu();
        onQueryChanged({
            ...dimensionQuery,
            valueFilter: getDefaultFilter(newFilterType)
        });
    };

    const handleFilterValueChanged = (newFilterValue: string | string[] | number) => {
        onQueryChanged({
            ...dimensionQuery,
            valueFilter: {
                type: dimensionQuery.valueFilter.type,
                query: newFilterValue
            }
        });
    }

    // MUI Switch onChange requires type (x: Y) => {} instead of (x: Y) => void
    const onChangeMUIWrapper = (newQuery: IDimensionQuery) => {
        onQueryChanged(newQuery);
        return {}
    }

    let filterComponent = null;
    let selectedValues = null;

    switch (dimensionQuery.valueFilter.type) {
        case FilterType.Item:
            if (dimensionQuery?.valueFilter?.query && dimension?.values) {
                const stringArray = dimensionQuery.valueFilter.query as string[];
                selectedValues = stringArray.map(code => dimension.values.find(o => o.code === code));
            }
            filterComponent =
                <ManualPickDimensionSelection
                    options={dimension.values}
                    selectedValues={selectedValues}
                    onQueryChanged={handleFilterValueChanged}
                />
            break;
        case FilterType.All:
            filterComponent =
                <AllDimensionSelection />
            break;
        case FilterType.From:
            filterComponent =
                <StartingFromDimensionSelection
                    options={dimension.values}
                    startingCode={dimensionQuery.valueFilter.query as string}
                    onQueryChanged={handleFilterValueChanged}
                />
            break;
        case FilterType.Top:
            filterComponent =
                <TopNDimensionSelection
                    numberOfItems={dimensionQuery.valueFilter.query as number}
                    onNumberChanged={handleFilterValueChanged}
                />
            break;
    }

    return (
        <ComponentWrapper direction="column" spacing={2}>
            <SelectorWrapper direction="row">
                <IconButton aria-label={anchorEl ? t('variableSelect.closeMenu') : t('variableSelect.openMenu')} component="span" style={{ paddingLeft: "0px", marginRight: "12px" }} onClick={handleOpenMenuClick}>
                    <MenuIcon />
                </IconButton>
                {filterComponent}
            </SelectorWrapper>

            {
                dimensionQuery.valueFilter.type !== FilterType.Item ? (
                    <ResultList dimensionValues={dimension.values} resolvedDimensionValueCodes={resolvedDimensionValueCodes} />
                ) : null
            }

            <SelectabilitySwitch onChange={value => onChangeMUIWrapper({ ...dimensionQuery, selectable: value })} selected={dimensionQuery.selectable} />
            {
                dimensionQuery.selectable && <DefaultSelectableDimensionSelection dimensionCode={dimension.code} resolvedDimensionValueCodes={resolvedDimensionValueCodes} options={dimension.values} />
            }

            <Menu open={anchorEl != null} anchorEl={anchorEl} onClose={closeMenu}>
                {
                    (Object.values(FilterType) as Array<FilterType>).map(queryType => {
                        const selected = dimensionQuery.valueFilter.type === queryType;
                        return (
                            <MenuItem key={queryType} selected={selected} onClick={() => handleMenuClick(queryType)}>
                                {selected ?
                                    <><ListItemIcon><CheckIcon /></ListItemIcon>{t(queryTypeLabels[queryType])}</> :
                                    <ListItemText inset>{t(queryTypeLabels[queryType])}</ListItemText>
                                }
                            </MenuItem>
                        );
                    })
                }
            </Menu>
        </ComponentWrapper>
    );
}
export default DimensionSelection;