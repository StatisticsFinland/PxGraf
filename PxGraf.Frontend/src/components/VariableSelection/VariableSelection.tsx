import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';

import { Stack, ListItemIcon, ListItemText, IconButton, MenuItem, Menu } from '@mui/material';

import MenuIcon from '@mui/icons-material/Menu';
import CheckIcon from '@mui/icons-material/Check';

import {
    getDefaultFilter, queryTypeLabels,
} from 'utils/variableSelectionHelpers';
import SelectabilitySwitch from './SelectabilitySwitch';
import ResultList from './ResultList';
import ManualPickVariableSelection from './FilterComponents/ManualPickVariableSelection';
import AllVariableSelection from './FilterComponents/AllVariableSelection';
import StartingFromVariableSelection from './FilterComponents/StartingFromVariableSelection';
import TopNVariableSelection from './FilterComponents/TopNVariableSelection';
import styled from 'styled-components';
import { IVariable } from 'types/cubeMeta';
import { FilterType, IVariableQuery } from 'types/query';
import DefaultSelectableVariableSelection from './DefaultSelectableVariableSelection';

interface IVariableSelectionProps {
    variable: IVariable
    resolvedVariableValueCodes: string[]
    query: IVariableQuery
    onQueryChanged: (newVarQuery: IVariableQuery) => void
}

const SelectorWrapper = styled(Stack)`
    width: 100%;
    align-items: flex-start;
`;

const ComponentWrapper = styled(Stack)`
    justify-content: space-between;
    align-items: flex-start;
`;

export const VariableSelection: React.FC<IVariableSelectionProps> = ({ variable, resolvedVariableValueCodes, query, onQueryChanged }) => {
    const [anchorEl, setAnchorEl] = useState(null);
    const { t } = useTranslation();

    const handleOpenMenuClick = (event) => {
        setAnchorEl(event.currentTarget);
    };

    const closeMenu = () => {
        setAnchorEl(null);
    };

    const handleMenuClick = (newFilterType: FilterType) => {
        closeMenu();
        onQueryChanged({
            ...query,
            valueFilter: getDefaultFilter(newFilterType)
        });
    };

    const handleFilterValueChanged = (newFilterValue: string | string[] | number) => {
        onQueryChanged({
            ...query,
            valueFilter: {
                type: query.valueFilter.type,
                query: newFilterValue
            }
        });
    }

    // MUI Switch onChange requires type (x: Y) => {} instead of (x: Y) => void
    const onChangeMUIWrapper = (newQuery: IVariableQuery) => {
        onQueryChanged(newQuery);
        return {}
    }

    let filterComponent = null;

    let selectedValues = null;


    switch (query.valueFilter.type) {
        case FilterType.Item:
            if (query?.valueFilter?.query && variable?.values) {
                const stringArray = query.valueFilter.query as string[];
                selectedValues = stringArray.map(code => variable.values.find(o => o.code === code));
            }
            filterComponent =
                <ManualPickVariableSelection
                options={variable.values}
                selectedValues={selectedValues}
                onQueryChanged={handleFilterValueChanged}
                />
            break;
        case FilterType.All:
            filterComponent =
                <AllVariableSelection />
            break;
        case FilterType.From:
            filterComponent =
                <StartingFromVariableSelection
                options={variable.values}
                startingCode={query.valueFilter.query as string}
                onQueryChanged={handleFilterValueChanged}
                />
            break;
        case FilterType.Top:
            filterComponent =
                <TopNVariableSelection
                    numberOfItems={query.valueFilter.query as number}
                    onNumberChanged={handleFilterValueChanged}
                />
            break;
    }

    return (
        <ComponentWrapper direction="column" spacing={2}>
            <SelectorWrapper direction="row">
                <IconButton aria-label={anchorEl ? t('variableSelect.closeMenu') : t('variableSelect.openMenu')} component="span" style={{ paddingLeft: "0px", marginRight: "12px" }} onClick={handleOpenMenuClick}>
                    <MenuIcon />
                </IconButton>
                {filterComponent}
            </SelectorWrapper>

            {
                query.valueFilter.type !== FilterType.Item ? (
                    <ResultList variableValues={variable.values} resolvedVariableValueCodes={resolvedVariableValueCodes} />
                ) : null
            }

            <SelectabilitySwitch onChange={value => onChangeMUIWrapper({ ...query, selectable: value })} selected={query.selectable} />
            {
                query.selectable && <DefaultSelectableVariableSelection variableCode={variable.code} resolvedVariableValueCodes={resolvedVariableValueCodes} options={variable.values} />
            }

            <Menu open={anchorEl != null} anchorEl={anchorEl} onClose={closeMenu}>
                {
                    (Object.values(FilterType) as Array<FilterType>).map(queryType => {
                        const selected = query.valueFilter.type === queryType;
                        return (
                            <MenuItem key={queryType} selected={selected} onClick={() => handleMenuClick(queryType)}>
                                {selected ?
                                    <><ListItemIcon><CheckIcon /></ListItemIcon>{t(queryTypeLabels[queryType])}</> :
                                    <ListItemText inset>{t(queryTypeLabels[queryType])}</ListItemText>
                                }
                            </MenuItem>
                        );
                    })
                }
            </Menu>
        </ComponentWrapper>
    );
}
export default VariableSelection;