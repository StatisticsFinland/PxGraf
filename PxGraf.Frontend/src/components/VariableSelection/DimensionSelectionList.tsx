import React from 'react';
import { Typography, Accordion, AccordionSummary, AccordionDetails } from '@mui/material';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import DimensionSelection from './DimensionSelection';
import { IDimension } from 'types/cubeMeta';
import { Query } from 'types/query';
import InfoBubble from 'components/InfoBubble/InfoBubble';
import { useTranslation } from 'react-i18next';
import styled from 'styled-components';
import { useTheme } from '@mui/material/styles';
import { UiLanguageContext } from 'contexts/uiLanguageContext';
import { sortedDimensions } from 'utils/sortingHelpers';
import { EditorContext } from '../../contexts/editorContext';

interface DimensionSelectionListProps {
    dimensions: IDimension[],
    resolvedDimensionCodes: { [key: string]: string[] },
    query: Query
}

const TitleWrapper = styled.div`
    display: flex;
    padding-top: 16px;
    padding-left: 16px;
    padding-right: 16px;
    align-items: center;
`;

const StyledAccordionDetails = styled(AccordionDetails)`
    background-color: #f8f8f8;
`;

const StyledAccordion = styled(Accordion)`
    border-top: 1px solid #dcdcdc;
`;

/**
 * Component for defining dimension filters and selectable dimensions in @see {@link Editor}.
 * @param {IDimension[]} dimensions Dimensions for the table in question.
 * @param {{ [key: string]: string[] }} resolvedDimensionCodes Resolved dimension value codes.
 * @param {Query} query Dimension queries.
 */
export const DimensionSelectionList: React.FC<DimensionSelectionListProps> = ({ dimensions, resolvedDimensionCodes, query }) => {
    const { t } = useTranslation();
    const theme = useTheme();
    const { uiContentLanguage } = React.useContext(UiLanguageContext);
    const { setQuery } = React.useContext(EditorContext);

    const infoContent = (
        <>
            {t("infoText.selectableButton")}
            <br />
            <br />
            {t("infoText.selectableInput")}
        </>
    );

    const selectedValues = (code: string ) => {
        //Formats the text to show 0 if no values are selected
        return resolvedDimensionCodes?.[code] ? resolvedDimensionCodes[code].length : 0;
    };

    return (
        <>
            <TitleWrapper>
                <Typography variant="body2" sx={{ color: theme.palette.text.secondary }}>{t('variableSelect.title')}</Typography>
                <InfoBubble info={infoContent} ariaLabel={t('variableSelect.title')} id="mainContent" />
            </TitleWrapper>
            {sortedDimensions(dimensions).map(dimension => {
                return (
                    <StyledAccordion key={dimension.code} defaultExpanded={true}>
                        <AccordionSummary
                            expandIcon={<ExpandMoreIcon />}
                            aria-controls="panel1a-content"
                        >
                            <Typography variant="h2"><b>{dimension.name[uiContentLanguage] ?? dimension.code} {selectedValues(dimension.code)}/{dimension.values.length}</b></Typography>
                        </AccordionSummary>
                        <StyledAccordionDetails>
                            <DimensionSelection
                                dimension={dimension}
                                resolvedDimensionValueCodes={resolvedDimensionCodes?.[dimension.code]}
                                dimensionQuery={query[dimension.code]}
                                onQueryChanged={newQuery => setQuery({ ...query, [dimension.code]: newQuery })}
                            />
                        </StyledAccordionDetails>
                    </StyledAccordion>
                );
            })}
        </>
    );
}

export default DimensionSelectionList;