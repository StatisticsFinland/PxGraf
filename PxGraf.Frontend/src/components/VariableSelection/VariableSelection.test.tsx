import React from 'react';
import { render } from "@testing-library/react";
import { IDimension, EDimensionType } from "types/cubeMeta";
import { FilterType, IVariableQuery } from "types/query";
import VariableSelection from "./VariableSelection";
import UiLanguageContext from "contexts/uiLanguageContext";

const mockVariable: IDimension =
{
    code: "Vuosi",
    name: {
        fi: "Vuosi",
        sv: "Ã…r",
        en: "Year"
    },
    type: EDimensionType.Time,
    values: [
        {
            code: "2018",
            name: {
                fi: "2018",
                sv: "2018",
                en: "2018"
            },
            isVirtual: false
        },
        {
            code: "2019",
            name: {
                fi: "2019",
                sv: "2019",
                en: "2019"
            },
            isVirtual: false
        },
        {
            code: "2020",
            name: {
                fi: "2020",
                sv: "2020",
                en: "2020"
            },
            isVirtual: false
        },
        {
            code: "2021",
            name: {
                fi: "2021*",
                sv: "2021*",
                en: "2021*"
            },
            isVirtual: false
        }
    ]
}

const mockQuery: IVariableQuery = {
    valueFilter: {
        type: FilterType.Top,
        query: 4
    },
    selectable: false,
    virtualValueDefinitions: null
};

const setLanguage = jest.fn();
const language = 'fi';

const setLanguageTab = jest.fn();
const languageTab = 'fi';

const availableUiLanguages = ['fi', 'en', 'sv'];
const uiContentLanguage = "fi";
const setUiContentLanguage = jest.fn();

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
        <UiLanguageContext.Provider value={{ language, setLanguage, languageTab, setLanguageTab, availableUiLanguages, uiContentLanguage, setUiContentLanguage }}>
            <VariableSelection
                variable={mockVariable}
                resolvedVariableValueCodes={["2018", "2019", "2020", "2021*"]}
                query={mockQuery}
                onQueryChanged={(newValues) => {}}/>
        </UiLanguageContext.Provider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});