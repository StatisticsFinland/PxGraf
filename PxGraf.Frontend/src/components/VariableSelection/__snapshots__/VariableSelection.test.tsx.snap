// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root sc-kkmypM kJaLMI css-1sazv7p-MuiStack-root"
  >
    <div
      class="MuiStack-root sc-dnaUSb dlJuVl css-m69qwo-MuiStack-root"
    >
      <span
        aria-label="variableSelect.openMenu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        role="button"
        style="padding-left: 0px; margin-right: 12px;"
        tabindex="0"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root sc-iCZwEW hwqTJh css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r0:"
          id=":r0:-label"
        >
          variableSelect.topFilter
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            inputmode="numeric"
            pattern="[0-9]*"
            type="text"
            value="4"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                variableSelect.topFilter
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiStack-root sc-dsLQwm bQNaJd css-nen11g-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        general.results:
      </p>
      <ul
        class="MuiList-root MuiList-padding MuiList-dense sc-braxZu kZeA-dF css-1f8gdww-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gJhJTp iUWrBn css-6lnq02-MuiListItem-root"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              2018
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gJhJTp iUWrBn css-6lnq02-MuiListItem-root"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              2019
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gJhJTp iUWrBn css-6lnq02-MuiListItem-root"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              2020
            </span>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding sc-gJhJTp iUWrBn css-6lnq02-MuiListItem-root"
        >
          <div
            class="MuiListItemText-root MuiListItemText-dense css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
            >
              2021*
            </span>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
        >
          <span
            class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-byenzh-MuiButtonBase-root-MuiSwitch-switchBase"
          >
            <input
              class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
              type="checkbox"
            />
            <span
              class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          variableSelect.selectable
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
